11850 gears notification api

javitonino opened the pull request:
Closes #11850 

This has a commit from #11779 (for notification model constants). Should be a clean merge.

- Adds two methods to manage notifications `create_notification` and `send_notification` in the Gears API.
- Moves instantiation of Value objects to a helper in the model `self.from_model`, marked as private API. This can be reverted (isolated commit). Rationale was that we were duplicating logic to instantiate an organization value in multiple places (user, notification). I changed all the models for consistency.

juanignaciosl commented on gears/carto_gears_api/lib/carto_gears_api/notifications/notification.rb:
```diff
@@ -0,0 +1,28 @@
+          id: notification.id,
+          body: notification.body,
+          icon: notification.icon,
+          organization: notification.organization,
```
This shouldn't be a `Carto::Organization` but a `CartoGearsApi::Organizations::Organization`.

juanignaciosl commented on gears/carto_gears_api/lib/carto_gears_api/notifications/notifications_service.rb:
```diff
@@ -0,0 +1,36 @@
+      # @see send_notification
+      #
+      # @param [String] body markdown body of the notification
+      # @param [String] icon name of the icon (defaults to +success+)
```
It's a little cumbersome, but maybe `CartoGearsApi::Notifications` module should expose the valid icons. Otherwise, we're coupling the model constants.

javitonino replied:
Yes, I thought about this last night. Couple of ideas:
- `ICONS = Carto::Notification::ICON` so the constant is directly exposed, and we do not have to do anything to keep in sync. But we only expose the list, we do not expose individual constants.
- Just copy the constant block and add a comment to `Carto::Notification` to suggest where it should be synced.
- Use a module for constants, and include it in both models.

Any preference? I think the most straight-forward is the second one. It adds a bit of overhead but I think adding constants is an infrequent enough change that it should be ok.

juanignaciosl replied:
I think that the most straightforward and safe approach is the first one (`ICONS = Carto::Notification::ICONS`, I guess that you missed the right-hand plural).

It's safe because we don't have to remember to update both sides, and it can't be easier and more clear. And it doesn't actually couple the API clients to `Carto::`-anything, because the right side is the implementation of the constant (just like the API methods contain calls to `Carto::` classes.

javitonino replied:
I like that option, but from the perspective of the user, we do not provide constants for each specific icon. They have it very easy to see the list of accepted values, though (just print the constant) and can manually create constants if they need them.

From our part, it is of course much easier to maintain (no maintenance needed), and we ensure this is always correct. 👍

juanignaciosl replied:
Oh, no. I spoke too fast. I'd do this, then:

```ruby
ICON_A = Carto::Notification::ICON_A
ICON_B = Carto::Notification::ICON_B
ICONS = [ICON_A, ICON_B]
```

I don't want users to need checking the code or hardcoding constants.

javitonino commented on gears/carto_gears_api/lib/carto_gears_api/notifications/notification.rb:
```diff
@@ -0,0 +1,32 @@
+      ICON_WARNING = Carto::Notification::ICON_WARNING
+      ICON_SUCCESS = Carto::Notification::ICON_SUCCESS
+      # List of possible icons
+      ICONS = Carto::Notification::ICONS
```
Heads-up for @matallo (as this is unlikely to show up as a merge conflict). We need to update this file when merging the second part of org notifications UI, since you touched the ICONS constants in that one if I am not mistaken.

matallo replied:
👍

javitonino commented:
Changes since last time:
- Copy `ICONS` constants to gears model
- Notification.organization has the correct type
- Move some `require` around. Previous version only worked if you always loaded the service, but it did not work if you just loaded the model.

juanignaciosl commented on gears/carto_gears_api/lib/carto_gears_api/notifications/notification.rb:
```diff
@@ -0,0 +1,33 @@
+      ICON_WARNING = Carto::Notification::ICON_WARNING
+      ICON_SUCCESS = Carto::Notification::ICON_SUCCESS
+      # List of possible icons
+      ICONS = Carto::Notification::ICONS
```
This should be `ICONS = [ICON_WARNING, ICON_SUCCESS]`, isn't it?

javitonino replied:
🤔 On second thought, why have this constant at all here? It's not used by our code, and I think the individual constants are more useful for API users.

juanignaciosl replied:
Oh, I didn't realize about that. If it's not used, 🔥 . Easy addition if needed.

javitonino commented:
Removed the constant and 🍏

juanignaciosl commented:
I must admit that sending a notification without sender at the console felt weird. Anyway, there's no explicit requirement for that, and only organization notifications would have a human sender, who would always be the same user, so currently it doesn't make sense. Merging and deploying 🚀
